pkg/ingester/ingester_activity.go

```diff
@@ -150,5 +150,76 @@ func (i *ActivityTrackerWrapper) ShutdownHandler(w http.ResponseWriter, r *http.
 func requestActivity(ctx context.Context, name string, req interface{}) string {
 	userID, _ := tenant.TenantID(ctx)
 	traceID, _ := tracing.ExtractSampledTraceID(ctx)
-	return fmt.Sprintf("%s: user=%q trace=%q request=%v", name, userID, traceID, req)
+
+	switch r := req.(type) {
+	case *client.QueryRequest:
+		// To minimize memory allocation, make use of an optimized stringer implementation
+		// for *client.QueryRequest type, as this request can be invoked multiple times per second.
+		return queryRequestActivity(name, userID, traceID, r)
+
+	default:
+		return fmt.Sprintf("%s: user=%q trace=%q request=%v", name, userID, traceID, req)
+	}
+}
+
+func queryRequestActivity(name, userID, traceID string, req *client.QueryRequest) string {
+	sb := bytes.NewBuffer(
+		make([]byte, 0, 8192),
+	)
+	sb.WriteString(name)
+	sb.WriteString(`: user="`)
+	sb.WriteString(userID)
+	sb.WriteString(`"`)
+	sb.WriteString(` trace="`)
+	sb.WriteString(traceID)
+	sb.WriteString(`"`)
+	sb.WriteString(" request=")
+	queryRequestToString(sb, req)
+
+	return sb.String()
+}
+
+func queryRequestToString(sb *bytes.Buffer, req *client.QueryRequest) {
+	if req == nil {
+		sb.WriteString("nil")
+		return
+	}
+	b := make([]byte, 0, 32)
+
+	sb.WriteString("&QueryRequest{")
+
+	sb.WriteString("StartTimestampMs:")
+	sb.Write(strconv.AppendInt(b, req.StartTimestampMs, 10))
+	sb.WriteString(",")
+
+	sb.WriteString("EndTimestampMs:")
+	sb.Write(strconv.AppendInt(b, req.EndTimestampMs, 10))
```

Comment:
Don't we need to reset `b` here. it still contains start timestamp.

Reply:
I was wondering why test didn't flag this. Reason is that we don't reassign `b` to the result of `strconv.AppendInt(b, req.StartTimestampMs, 10)`, so `b` is still empty slice after this operation, and can be reused. It means my comment was wrong, and we don't need the reset.